Open source basic Alert and Solar Notebooks

cyiallou opened the pull request.

cyiallou commented:
Instruction documents are still needed.

cyiallou commented:
Rebased to latest main branch

cyiallou committed:
Add instruction documents

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou commented:
Added the instruction docs.

cyiallou commented:
Replaced a hardcoded value with the corresponding variable (a remnant from testing)

cwasicki commented:
Couple of comments:
* The notebooks have execution errors. Better make sure that the committed versions are working out-of-the box and remove all cell outputs.
* I think we should only have a single notebook with both languages, maintaining two versions of the same notebook can easily go out of sync. If you feel it gets too bloated maybe we can move some text out of the notebook again? The original idea to have a lot of doc there originates from the time where we were planning to send self-contained notebooks via mail. What do you think?
* Is it possible to move more of the logic in the alerting notebook into the library?

cyiallou commented:
> The notebooks have execution errors. Better make sure that the committed versions are working out-of-the box and remove all cell outputs.

The notebooks do work out of the box. The execution errors you see are basically the Deepnote UI elements which display some variable values and for the deepnote plots.

> I think we should only have a single notebook with both languages, maintaining two versions of the same notebook can easily go out of sync. 

I agree.

> If you feel it gets too bloated maybe we can move some text out of the notebook again? The original idea to have a lot of doc there originates from the time where we were planning to send self-contained notebooks via mail. What do you think?

Personally, I prefer to have self-contained notebooks instead of having to read a separate document in order to understand how to execute it. On the other hand, I'd hate having to read a bunch of text too. 

> Is it possible to move more of the logic in the alerting notebook into the library?

I suppose so. For example, `generate_html_email` is a large function which I created so that the user can have an example template email (as a starting point for them to customise). It can be moved into the library. I'll check other parts too.


**Next steps:**
- I'll check if I can remove the deepnote cell outputs (the ones with execution error).
- Merge the notebooks into 1 for both languages.
- Move some of the logic out of the alerting notebook.

cyiallou committed:
Support grouping alerts by component in summaries and emails

- Added `group_by_component` flag to `generate_alert_summary` and
  `generate_alert_json` to enable per-component grouping.
- Updated `generate_alert_email` to support component-level grouping in
  the summary.
- Adjusted alert aggregation logic to dynamically determine grouping
  levels.
- Improved formatting of alert summaries to reflect component-specific
  details when enabled.

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Add severity-based row coloring in alert tables

Applied color-coding to alert table rows based on severity levels (error
and warning).

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Sort alerts by severity in email notifications

- Added `SEVERITY_ORDER` to define alert priority (error > warning > state).
- Introduced a `sort_by_severity` flag in `generate_alert_table` to
  optionally sort alerts.
- Updated `generate_alert_email` to pass the sorting flag for better
  visibility of critical alerts.

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Add alerts module for generating and formatting alert emails

- Added `alert_email.py` with functions to format alert summaries,
  tables, and full HTML emails.
- Implemented utilities for timestamp parsing, elapsed time calculations,
  and structured JSON output.
- Included CSS styling for email formatting.

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Open source basic Solar Maintenance Notebook

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Add release notes

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cwasicki commented on src/frequenz/lib/notebooks/alerts/alert_email.py:
```diff
@@ -0,0 +1,304 @@
+        )
+        .reset_index()
+    )
+
```
There could be an option to filter out groups without any error or warning.

cyiallou replied:
Good idea. By default or as an option for the user?

cwasicki replied:
Maybe as an option for the user but True by default. For some customer I could imagine that they would still be interested in seeing which components were analyzed.

cyiallou committed:
Filter out groups with no alerts in email summaries

- Added a `filter_no_alerts` option to `generate_alert_summary` to
  exclude groups with zero errors and warnings.
- Updated `generate_alert_email` to support this filtering for cleaner
  alert reports.
- Made `notebook_url` optional in `generate_alert_email`.
- Applied pylint disable for too many arguments in `generate_alert_email`.
  This should be improved in the future (perhas with a dataclass)

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Update alert_email module docstring

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>

cyiallou committed:
Open source basic Alert notebook

Signed-off-by: cyiallou - Costas <42914163+cyiallou@users.noreply.github.com>
